src/webserver/routing.rs

```diff
@@ -0,0 +1,496 @@
+    Execute(PathBuf),
+    NotFound(PathBuf),
+    Redirect(Uri),
+    Serve(PathBuf),
+}
+
+pub trait ExecutionStore {
+    async fn contains(&self, path: &PathBuf) -> bool;
+}
+
+pub trait RoutingConfig {
+    fn prefix(&self) -> &Uri;
+}
+
+pub async fn calculate_route<T, C>(uri: Uri, store: &T, config: &C) -> RoutingAction
+where
+    T: ExecutionStore,
+    C: RoutingConfig,
+{
+    match check_uri(&uri, config) {
+        Ok(path) => match path.clone().extension() {
+            None => calculate_route_without_extension(&uri, path, store).await,
+            Some(extension) => calculate_route_with_extension(path, extension, store).await,
+        },
+        Err(action) => action,
+    }
+}
+
+fn check_uri<C>(uri: &Uri, config: &C) -> Result<PathBuf, RoutingAction>
+where
+    C: RoutingConfig,
+{
+    if uri.path().starts_with(config.prefix().path()) {
+        let mut result = String::from("/");
+        result.push_str(
+            uri.path()
+                .strip_prefix(config.prefix().path())
+                .expect("Unable to remove expected prefix from path"),
+        );
+
+        Ok(PathBuf::from(result))
+    } else {
+        Err(Redirect(config.prefix().clone()))
+    }
+}
+
+async fn calculate_route_without_extension<T>(
+    uri: &Uri,
+    mut path: PathBuf,
+    store: &T,
+) -> RoutingAction
+where
+    T: ExecutionStore,
+{
+    if uri.path().ends_with(FORWARD_SLASH) {
+        path.push(INDEX);
+        find_execution_or_not_found(&path, store).await
+    } else {
+        let path_with_ext = path.with_extension(EXECUTION_EXTENSION);
+        match find_execution(&path_with_ext, store).await {
+            Some(action) => action,
+            None => Redirect(append_to_uri_path(&uri, FORWARD_SLASH)),
+        }
+    }
+}
+
+async fn calculate_route_with_extension<T>(
+    path: PathBuf,
+    extension: &OsStr,
+    store: &T,
+) -> RoutingAction
+where
+    T: ExecutionStore,
+{
+    if extension == EXECUTION_EXTENSION {
+        find_execution_or_not_found(&path, store).await
+    } else {
+        Serve(path)
+    }
+}
+
+async fn find_execution_or_not_found<T>(path: &PathBuf, store: &T) -> RoutingAction
+where
+    T: ExecutionStore,
+{
+    match find_execution(path, store).await {
+        None => find_not_found(path, store).await,
+        Some(execute) => execute,
+    }
+}
+
+async fn find_execution<T>(path: &PathBuf, store: &T) -> Option<RoutingAction>
+where
+    T: ExecutionStore,
+{
+    if store.contains(path).await {
+        Some(Execute(path.clone()))
+    } else {
+        None
+    }
+}
+
+async fn find_not_found<T>(path: &PathBuf, store: &T) -> RoutingAction
+where
+    T: ExecutionStore,
+{
+    let mut parent = path.parent();
+    while let Some(p) = parent {
+        let target = p.join(NOT_FOUND);
+        if store.contains(&target).await {
+            return NotFound(target);
+        } else {
+            parent = p.parent()
+        }
+    }
+
+    Error(path_to_string(path))
+}
```

Comment:
```suggestion
async fn find_not_found<T>(mut path: &Path, store: &T) -> RoutingAction
where
    T: ExecutionStore,
{
    while let Some(parent) = path.parent() {
        let target = parent.join(NOT_FOUND);
        if store.contains(&target).await {
            return NotFound(target);
        }
        path = parent;
    }

    Error(path_to_string(path))
}

```

Reply:
good catch, I usually avoid early returns so am trying to use them (and forgetting the consequences!)